File: NalabsNativeLib/TextAnalyzer.cs
using Mdu.Nalabs.Metrics;

namespace Addiva.Nalabs.Interop;

public class TextAnalyzer
{
    private static readonly ConjunctionsMetric ConjunctionsMetric = new();
    private static readonly NVMetric NVMetric = new();
    private static readonly OptionalityMetric OptionalityMetric = new();
    private static readonly SubjectivityMetric SubjectivityMetric = new();
    private static readonly ReferencesMetric ReferencesMetric = new();
    private static readonly WeaknessMetric WeaknessMetric = new();
    private static readonly ImperativesMetric ImperativesMetric = new();
    private static readonly ContinuancesMetric ContinuancesMetric = new();
    private static readonly Imperatives2 Imperatives2Metric = new();
    private static readonly References2 References2Metric = new();

    public static Addiva.Nalabs.Interop.TextAnalysis AnalyzeText(string text)
    {
        var analysis = new Addiva.Nalabs.Interop.TextAnalysis
        {
            ARI = CalculateARIScore(text),
            Conjunctions = ConjunctionsMetric.Analyze(text),
            VaguePhrases = NVMetric.Analyze(text),
            Optionality = OptionalityMetric.Analyze(text),
            Subjectivity = SubjectivityMetric.Analyze(text),
            References = ReferencesMetric.Analyze(text),
            Weakness = WeaknessMetric.Analyze(text),
            Imperatives = ImperativesMetric.Analyze(text),
            Continuances = ContinuancesMetric.Analyze(text),
            Imperatives2 = Imperatives2Metric.Analyze(text),
            References2 = References2Metric.Analyze(text)
        };

        return analysis;
    }

    private static double CalculateARIScore(string text)
    {
        var sentences = text
            .Trim()
            .Split(['.'])
            .Where(n => !string.IsNullOrEmpty(n) && n != " " && n.Length > 1)
            .ToArray();

        var words = text
            .Trim()
            .Split([' ', '\n'])
            .Where(n => !string.IsNullOrEmpty(n) && n != " ")
            .ToArray();

        double sum = 0;

        foreach (var s in sentences)
        {
            sum += s.Split(' ').Length;
        }

        double wordSum = 0;

        foreach (var s in words)
        {
            wordSum += s.Length;
        }

        return sum / sentences.Length + 9 * (wordSum / words.Count());
    }
}

File: nalabs_plugin/Nalabs/src/nalabs/core/TextAnalyzer.java
package nalabs.core;

import java.util.Arrays;
import java.util.stream.Collectors;

public class TextAnalyzer
{
    private static ConjunctionsMetric ConjunctionsMetric = new ConjunctionsMetric();
    private static NVMetric NVMetric = new NVMetric();
    private static OptionalityMetric OptionalityMetric = new OptionalityMetric();
    private static SubjectivityMetric SubjectivityMetric = new SubjectivityMetric();
    private static ReferencesMetric ReferencesMetric = new ReferencesMetric();
    private static WeaknessMetric WeaknessMetric = new WeaknessMetric();
    private static ImperativesMetric ImperativesMetric = new ImperativesMetric();
    private static ContinuancesMetric ContinuancesMetric = new ContinuancesMetric();
    private static Imperatives2 Imperatives2Metric = new Imperatives2();
    private static References2 References2Metric = new References2();

    public static nalabs.core.TextAnalysis AnalyzeText(String text) {
    	nalabs.core.TextAnalysis analysis = new nalabs.core.TextAnalysis() {
    		{
	            ARI = CalculateARIScore(text);
	            conjunctions = ConjunctionsMetric.analyze(text);
	            vaguePhrases = NVMetric.analyze(text);
	            optionality = OptionalityMetric.analyze(text);
	            subjectivity = SubjectivityMetric.analyze(text);
	            references = ReferencesMetric.analyze(text);
	            weakness = WeaknessMetric.analyze(text);
	            imperatives = ImperativesMetric.analyze(text);
	            continuances = ContinuancesMetric.analyze(text);
	            imperatives2 = Imperatives2Metric.analyze(text);
	            references2 = References2Metric.analyze(text);
    		}
        };

        return analysis;
    }

    private static double CalculateARIScore(String text)
    {
        String[] sentences = (String[])Arrays.stream(text.trim().split("\\."))
        		.filter(n -> !n.isEmpty() && n != " " && n.length() > 1)
        		.toArray(String[]::new);     

        String[] words = (String[])Arrays.stream(text.trim().split(" |\n"))
            .filter(n -> !n.isEmpty() && n != " ")
            .toArray(String[]::new);

        double sum = 0;

        for (String s : sentences)
        {
            sum += s.split(" ").length;
        }

        double wordSum = 0;

        for (String s : words)
        {
            wordSum += s.length();
        }

        return sum / sentences.length + 9 * (wordSum / words.length);
    }
}